### PR TITLE
TDL-18662: Dict based to class based refactoring

### PR DESCRIPTION
**Pull request by karanpanchal-crest**
# Description of change
 - Implement  dict based to class-based

# Manual QA steps
 - Compare all catalogs with new catalog.
 - Compare old catalog (with state and without state) with new catalog.
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_snapchat_ads/streams.py [hunk 1]
```diff
@@ -17,758 +17,1419 @@
...
+            state['bookmarks'][stream] = {}
+
+        state['bookmarks'][stream][key] = value
+        LOGGER.info('Write state for Stream: {}, {} ID: {}, value: {}'.format(
+            stream, parent, parent_id, value))
+        singer.write_state(state)
+
+    # To transform string to datetime 
+    def transform_datetime(self, this_dttm):
+        import pdb; pdb.set_trace()
```
Comment (harshpatel4crest) [inline tap_snapchat_ads/streams.py:117]: Please remove this if not required.

File: tap_snapchat_ads/streams.py [hunk 2]
```diff
@@ -17,758 +17,1419 @@
...
+    bookmark_query_field_from = None
+    bookmark_query_field_to = None
+    date_window_size = 1
+    targeting_country_ind = False
+    targeting_group = None
+    targeting_type = None
+    children = []
+
+    # To write schema in output
+    def write_schema(self, catalog, stream_name):
```
Comment (savan-chovatiya) [inline tap_snapchat_ads/streams.py:57]: Add function comments for all the added functions.

File: tap_snapchat_ads/streams.py [hunk 3]
```diff
@@ -17,758 +17,1419 @@
...
+                            max_bookmark_value = last_datetime
+
+                        max_bookmark_dttm = strptime_to_utc(max_bookmark_value)
+
+                        if bookmark_dttm > max_bookmark_dttm:
+                            max_bookmark_value = strftime(bookmark_dttm)
+
+                        # Keep only records whose bookmark is after the last_datetime
+                        if bookmark_dttm >= last_dttm:
+                            # LOGGER.info('record1: {}'.format(record)) # TESTING, comment out
```
Comment (savan-chovatiya) [inline tap_snapchat_ads/streams.py:159]: These kinds of comments are present in multiple places. Remove those comments.

File: tap_snapchat_ads/streams.py [hunk 4]
```diff
@@ -17,758 +17,1419 @@
...
+    bookmark_query_field_from = 'start_time'
+    bookmark_query_field_to = 'end_time'
+    path= 'adaccounts/{parent_id}/stats'
+    data_key_array = 'timeseries_stats'
+    data_key_record = 'timeseries_stat'
+    date_window_size = 30
+    paging = False
+    parent = 'ad_account'
+    params = {
+        # 'test': 'true', # COMMENT OUT
```
Comment (savan-chovatiya) [inline tap_snapchat_ads/streams.py:661]: These kinds of comments are present already but let's remove them as part of refactoring as it looks like debugging comment.

**Review by savan-chovatiya — CHANGES_REQUESTED** [review]

File: tap_snapchat_ads/streams.py (hunk 1 above)
Comment (karanpanchal-crest) [reply tap_snapchat_ads/streams.py]: Removed unnecessary comments.

File: tap_snapchat_ads/streams.py (hunk 2 above)
Comment (karanpanchal-crest) [reply tap_snapchat_ads/streams.py:57]: Added doc string for all newly added functions

File: tap_snapchat_ads/streams.py (hunk 3 above)
Comment (karanpanchal-crest) [reply tap_snapchat_ads/streams.py]: Removed unnecessary comments.

File: tap_snapchat_ads/streams.py (hunk 4 above)
Comment (karanpanchal-crest) [reply tap_snapchat_ads/streams.py]: Removed unnecessary comments.

**Review by savan-chovatiya — APPROVED** [review]

**Review by dbshah1212 — APPROVED** [review]

**Review by kspeer825 — APPROVED** [review]
Assuming this will not be merged without updated tests from other PR.

**Review by KrisPersonal — APPROVED** [review]

**Review by RushiT0122 — APPROVED** [review]

**Pull request merged**